File: NC-150/03_Sliding_Window/16_longest_substring_no_repeat.py
"""
LeetCode 3: Longest Substring Without Repeating Characters
https://leetcode.com/problems/longest-substring-without-repeating-characters/

Given a string s, find the length of the longest substring without repeating characters.

Time Complexity: O(n)
Space Complexity: O(min(m, n)) where m is charset size
"""


class Solution:
    # Solution 1: Sliding Window with Set
    def lengthOfLongestSubstring(self, s: str) -> int:
        chars = set()
        left = 0
        max_len = 0
        
        for right in range(len(s)):
            # Shrink window until no duplicate
            while s[right] in chars:
                chars.remove(s[left])
                left += 1
            
            chars.add(s[right])
            max_len = max(max_len, right - left + 1)
        
        return max_len
    
    # Solution 2: Sliding Window with HashMap (optimized)
    def lengthOfLongestSubstringOptimized(self, s: str) -> int:
        last_index = {}
        left = 0
        max_len = 0
        
        for right in range(len(s)):
            if s[right] in last_index and last_index[s[right]] >= left:
                left = last_index[s[right]] + 1
            
            last_index[s[right]] = right
            max_len = max(max_len, right - left + 1)
        
        return max_len
    
    # Solution 3: Pythonic with enumerate
    def lengthOfLongestSubstringPythonic(self, s: str) -> int:
        last_seen = {}
        start = 0
        max_len = 0
        
        for i, char in enumerate(s):
            if char in last_seen and last_seen[char] >= start:
                start = last_seen[char] + 1
            else:
                max_len = max(max_len, i - start + 1)
            
            last_seen[char] = i
        
        return max_len


if __name__ == "__main__":
    sol = Solution()
    
    # Test Case 1: Basic case
    print(f"Test 1 'abcabcbb': {sol.lengthOfLongestSubstring('abcabcbb')}")
    # Expected: 3 ("abc")
    
    # Test Case 2: All same characters
    print(f"Test 2 'bbbbb': {sol.lengthOfLongestSubstring('bbbbb')}")
    # Expected: 1 ("b")
    
    # Test Case 3: All unique
    print(f"Test 3 'pwwkew': {sol.lengthOfLongestSubstring('pwwkew')}")
    # Expected: 3 ("wke")
    
    # Test Case 4: Empty string
    print(f"Test 4 '': {sol.lengthOfLongestSubstring('')}")
    # Expected: 0
    
    # Test Case 5: Single character
    print(f"Test 5 'a': {sol.lengthOfLongestSubstring('a')}")
    # Expected: 1
    
    # Test Case 6: Two different characters
    print(f"Test 6 'au': {sol.lengthOfLongestSubstring('au')}")
    # Expected: 2
    
    # Test Case 7: Space in string
    print(f"Test 7 'a b c': {sol.lengthOfLongestSubstring('a b c')}")
    # Expected: 3
    
    # Test Case 8: Repeating pattern
    print(f"Test 8 'abba': {sol.lengthOfLongestSubstring('abba')}")
    # Expected: 2

File: NC-150/03_Sliding_Window/16_longest_substring_no_repeat.cpp
/*
 * LeetCode 3: Longest Substring Without Repeating Characters
 * https://leetcode.com/problems/longest-substring-without-repeating-characters/
 * 
 * Given a string s, find the length of the longest substring without repeating characters.
 * 
 * Time Complexity: O(n)
 * Space Complexity: O(min(m, n)) where m is charset size
 */

#include <iostream>
#include <string>
#include <unordered_set>
#include <unordered_map>
#include <algorithm>
using namespace std;

class Solution {
public:
    // Solution 1: Sliding Window with Set
    int lengthOfLongestSubstring(string s) {
        unordered_set<char> chars;
        int left = 0;
        int maxLen = 0;
        
        for (int right = 0; right < s.length(); right++) {
            // Shrink window until no duplicate
            while (chars.count(s[right])) {
                chars.erase(s[left]);
                left++;
            }
            
            chars.insert(s[right]);
            maxLen = max(maxLen, right - left + 1);
        }
        
        return maxLen;
    }
    
    // Solution 2: Sliding Window with HashMap (optimized)
    // Jump left pointer directly to avoid incremental moves
    int lengthOfLongestSubstringOptimized(string s) {
        unordered_map<char, int> lastIndex;
        int left = 0;
        int maxLen = 0;
        
        for (int right = 0; right < s.length(); right++) {
            if (lastIndex.count(s[right]) && lastIndex[s[right]] >= left) {
                left = lastIndex[s[right]] + 1;
            }
            
            lastIndex[s[right]] = right;
            maxLen = max(maxLen, right - left + 1);
        }
        
        return maxLen;
    }
    
    // Solution 3: Array instead of HashMap (for ASCII)
    int lengthOfLongestSubstringArray(string s) {
        int lastIndex[128];
        fill(lastIndex, lastIndex + 128, -1);
        
        int left = 0;
        int maxLen = 0;
        
        for (int right = 0; right < s.length(); right++) {
            if (lastIndex[s[right]] >= left) {
                left = lastIndex[s[right]] + 1;
            }
            
            lastIndex[s[right]] = right;
            maxLen = max(maxLen, right - left + 1);
        }
        
        return maxLen;
    }
};

int main() {
    Solution sol;
    
    // Test Case 1: Basic case
    cout << "Test 1 'abcabcbb': " << sol.lengthOfLongestSubstring("abcabcbb") << endl;
    // Expected: 3 ("abc")
    
    // Test Case 2: All same characters
    cout << "Test 2 'bbbbb': " << sol.lengthOfLongestSubstring("bbbbb") << endl;
    // Expected: 1 ("b")
    
    // Test Case 3: All unique
    cout << "Test 3 'pwwkew': " << sol.lengthOfLongestSubstring("pwwkew") << endl;
    // Expected: 3 ("wke")
    
    // Test Case 4: Empty string
    cout << "Test 4 '': " << sol.lengthOfLongestSubstring("") << endl;
    // Expected: 0
    
    // Test Case 5: Single character
    cout << "Test 5 'a': " << sol.lengthOfLongestSubstring("a") << endl;
    // Expected: 1
    
    // Test Case 6: Two different characters
    cout << "Test 6 'au': " << sol.lengthOfLongestSubstring("au") << endl;
    // Expected: 2
    
    // Test Case 7: Space in string
    cout << "Test 7 'a b c': " << sol.lengthOfLongestSubstring("a b c") << endl;
    // Expected: 3 ("a b" or " bc")
    
    // Test Case 8: Repeating pattern
    cout << "Test 8 'abba': " << sol.lengthOfLongestSubstring("abba") << endl;
    // Expected: 2 ("ab" or "ba")
    
    return 0;
}

